### PR TITLE
#371: fix missing systemTimestamp in processor-emitted events

### LINKED ISSUE #371: Events created by EventsourcedProcessor lack system timestamp
The method `EventsourcedProcessor.createEvent` does not set the `systemTimestamp` on the `DurableEvent` it creates, therefore it is always `0L` when reading it.

### PR DESCRIPTION
**Pull request by danbim**
Fixes #371

### CODE REVIEW

**Comment by krasserm** [comment]
In a later version of Eventuate, there will be a dedicated event processing protocol (instead of abusing the replication protocol) for handling issues like these more specifically

**Comment by danbim** [comment]
Sounds good. I'll update the PR later.

File: eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/log/EventLog.scala [hunk 1]
```diff
@@ -596,7 +596,8 @@ abstract class EventLog[A <: EventLogState](id: String) extends Actor with Event
       case (acc, e) =>
         snr += 1
 
-        val e2 = e.prepare(id, snr, e.systemTimestamp)
+        val timestamp = if (e.systemTimestamp == 0L) currentSystemTime else e.systemTimestamp
```
Comment (krasserm) [inline eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/log/EventLog.scala]: Pass the current system time via a `systemTimestamp` parameter as done in `prepareEvents`. All events of an atomically written batch should have the same system time.

**Comment by krasserm** [comment]
And please write tests for all changes/additions you make.

**Comment by danbim** [comment]
Happy New Year :) I've added tests for the EventLog-based implementation. I hope these make sense now. As a big part of the EventLogSpec relies on `systemTimestamp = 0L` to be interpreted as the current system timestamp (in TestEventLog) I've had to "turn around this assumption" for this test case in which we want to assure that the timestamp is overridden if it is set to `0L`.

I have no test yet to actually verify that an `EventsourcedProcessor` now contains a correct system timestamp and I'm not sure if this should be done in the first place as it should be able to rely on the EventLog implementation conforming to its expected behaviour. WDYT?

File: eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import com.rbmhtechnology.eventuate.ReplicationProtocol.ReplicationWrite
+import com.rbmhtechnology.eventuate.utilities.timeoutDuration
+import com.rbmhtechnology.eventuate.DurableEvent
+import com.rbmhtechnology.eventuate.VectorTime
+
+trait EventLogSystemTimeSpec extends TestKitBase with EventLogSpecSupport {
+
+  import EventLogSpec._
+
+  val dl = system.deadLetters
```
Comment (krasserm) [inline eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala]: Not used, can be removed.

File: eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  implicit val implicitTimeout = Timeout(timeoutDuration)
```
Comment (krasserm) [inline eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala]: Not used, can be removed.

File: eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  val SystemTimestamp = 123L
+
+  "An event log" must {
+    "update an event's system timestamp" in {
+      log ! Write(List(event("a").copy(systemTimestamp = 3L)), system.deadLetters, replyToProbe.ref, 0, 0)
+      replyToProbe.expectMsgType[WriteSuccess].events.head.systemTimestamp should be(SystemTimestamp)
+    }
+    "update a replicated events system timestamp if not defined (systemTimestamp == 0L, cf. #371)" in {
+      val evt = DurableEvent("a", emitterIdA, processId = logId, vectorTimestamp = VectorTime(remoteLogId -> 1L))
+      registerCollaborator(aggregateId = None, collaborator = replyToProbe)
```
Comment (krasserm) [inline eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala]: Don't use `replyToProbe` here as you're not expecting a reply message. Instead use 

```scala
val collaborator = registerCollaborator()
```

to create a `TestProbe`.

File: eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      log ! ReplicationWrite(List(evt), Map(remoteLogId -> ReplicationMetadata(5, VectorTime.Zero)))
+      replyToProbe.expectMsgType[Written].event.systemTimestamp should not be UndefinedTimestamp
+    }
+    "not update a replicated events system timestamp if defined (systemTimestamp != 0L, cf. #371)" in {
+      val evt = DurableEvent("a", emitterIdA, processId = logId, vectorTimestamp = VectorTime(remoteLogId -> 1L), systemTimestamp = SystemTimestamp)
+      registerCollaborator(aggregateId = None, collaborator = replyToProbe)
```
Comment (krasserm) [inline eventuate-core/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpec.scala]: See previous comment

File: eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecLeveldb.scala [hunk 6]
```diff
@@ -32,13 +32,18 @@ trait LocationCleanupLeveldb extends LocationCleanup {
...
-      val logProps = Props(new TestEventLog(logId)).withDispatcher("eventuate.log.dispatchers.write-dispatcher")
+    def props(logId: String, batching: Boolean, currentSystemTimeOpt: Option[Long] = None): Props = {
+      val logProps = Props(new TestEventLog(logId, currentSystemTimeOpt))
+        .withDispatcher("eventuate.log.dispatchers.write-dispatcher")
       if (batching) Props(new BatchingLayer(logProps)) else logProps
     }
   }
 
-  class TestEventLog(id: String) extends LeveldbEventLog(id, "log-test") with SingleLocationSpec.TestEventLog[LeveldbEventLogState] {
+  class TestEventLog(id: String, currentSystemTimeOpt: Option[Long] = None) extends LeveldbEventLog(id, "log-test")
```
Comment (krasserm) [inline eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecLeveldb.scala]: Can be simplified:

```scala
class TestEventLog(id: String, override val currentSystemTime: Long)
```

There's no need to work with `Option` here.

File: eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecLeveldb.scala [hunk 7]
```diff
@@ -32,13 +32,18 @@ trait LocationCleanupLeveldb extends LocationCleanup {
...
+    with SingleLocationSpec.TestEventLog[LeveldbEventLogState] {
+
+    override def currentSystemTime: Long = currentSystemTimeOpt.getOrElse(0L)
```
Comment (krasserm) [inline eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecLeveldb.scala]: Must be removed

File: eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecLeveldb.scala [hunk 8]
```diff
@@ -61,7 +66,9 @@ trait SingleLocationSpecLeveldb extends SingleLocationSpec with LocationCleanupL
     _log
 
   def logProps(logId: String): Props =
-    RestarterActor.props(TestEventLog.props(logId, batching))
+    RestarterActor.props(TestEventLog.props(logId, batching, currentSystemTimeOpt))
+
+  def currentSystemTimeOpt: Option[Long] = None
```
Comment (krasserm) [inline eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecLeveldb.scala]: Define

```scala
def currentSystemTime: Long = 0L
```

in trait `SingleLocationSpec`.

File: eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpecLeveldb.scala [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import akka.actor.ActorSystem
+import akka.testkit.TestKit
+import com.rbmhtechnology.eventuate.SingleLocationSpecLeveldb
+
+class EventLogSystemTimeSpecLeveldb extends TestKit(ActorSystem("test", EventLogSpecLeveldb.config))
+  with EventLogSystemTimeSpec
+  with SingleLocationSpecLeveldb {
+
+  override def currentSystemTimeOpt: Option[Long] = Some(SystemTimestamp)
```
Comment (krasserm) [inline eventuate-log-leveldb/src/it/scala/com/rbmhtechnology/eventuate/log/EventLogSystemTimeSpecLeveldb.scala]: Change to 

```scala
override def currentSystemTime: Long = SystemTimestamp
```

File: eventuate-log-cassandra/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecCassandra.scala [hunk 10]
```diff
@@ -136,6 +136,8 @@ trait SingleLocationSpecCassandra extends SingleLocationSpec with LocationCleanu
 
   private def _createLog(failureSpec: TestFailureSpec, indexProbe: ActorRef): ActorRef =
     system.actorOf(logProps(logId, failureSpec, indexProbe))
+
+  def currentSystemTimeOpt: Option[Long] = None
```
Comment (krasserm) [inline eventuate-log-cassandra/src/it/scala/com/rbmhtechnology/eventuate/LocationSpecCassandra.scala]: See comments on `LocationSpecLeveldb`

**Pull request merged**